File: tools/filter_allocs.js
//outputs memory allocations and memory accesses that have alloc-location information
//useful for focusing on references to dynamically allocated memory
//
//rough benchmarks put the JS version at twice as fast as python (wow!)
//requires npm install split

//usage: node filter_allocs.js < file.gz > output.json

var zlib = require('zlib');
var split = require('split');
var fs = require('fs');

//var inp = fs.createReadStream('../test/wiredtiger/ldb-readrandom-24t.json.gz');
var inp = process.stdin;

inp.pipe(zlib.createGunzip()).pipe(split()).on('data', function(line){
	e = JSON.parse(line)
	//console.log(line)
	//console.log(obj)
	//console.log("-----")

	if(e['event'] == "allocation"){
		console.log(line);
	}
	if(e['event'] == 'memory-access'){
		if(e['alloc-location'] != "<unknown>"){
			console.log(line);
			process.exit(0);
		}
	}
})

File: tools/filter_allocs.py
#!/usr/local/bin/python
import gzip, sys, json

#f = open("../test/wiredtiger/wt-ldb-readrand2.json.gz")
f = open("../test/wiredtiger/ldb-readrandom-24t.json.gz")
gzip_file = gzip.GzipFile(fileobj=f)

#print dir(sys)
#print dir(gzip_file)
#print gzip_file.__sizeof__()
#sys.exit()

#allocs = dict()
count = 0
for line in gzip_file:
	#if count == 10:
	#	sys.exit()
	#else:
	#	count = count + 1

	e = json.loads(line)
	if e['event'] == "allocation":
		#if e['alloc-base'] in allocs:
		#	print "free " + json.dumps(e)
		#else:
		#allocs[e['alloc-base']] = 1
		#print json.dumps(e)
		print line.strip()
	if e['event'] == "memory-access":
		if e['alloc-location'] != '<unknown>':
			#print json.dumps(e)
			print gzip_file.fileobj.tell()
			print line
			sys.exit()
	else:
		pass